### ISSUE TITLE
Storage reclaim overshooting

### ISSUE THREAD

**Issue by bkchr** [paritytech/polkadot-sdk#6020]
Storage reclaim is returning unused storage proof space to the on chain logic to enable follow up transactions to use this storage proof space. The reclaim is done by requesting the storage proof size on the node side and then correcting the estimated storage proof size in the runtime. It was detected that a chain was able to overshoot the total PoV budget while using storage reclaim. 

The problem is that storage reclaim doesn't take the storage root calculation into account. For items that are deleted or added, there is not happening any kind of trie access as these are just noted in the overlay. However, when calculating the storage root the item is looked up in the trie to add new nodes/delete/merge nodes. This lookup ultimately leads to using much more storage proof size that anticipated (depending on the state size/layout).

@skunert proposed that when adding/deleting a storage item, the key is looked up in the state to emulate what the storage root calculation is doing. This is working kind of good, but still misses to account for some nodes that are for example loaded when merging two nodes. However, it can work as a first solution to reduce the impact of the issue. A proper solution is to incrementally calculate the storage root by inserting/deleting the storage items into a memory DB as done by the storage root calculation. This still brings problems with transient storage items (items that are only available in a block) as they will lead to accessing the trie, but this can not be solved until there are proper host functions for these transient storage items.

**Comment by bkchr** [comment]
(Benchmarking itself also doesn't take this into account, but it generally is probably overshooting enough so that this doesn't lead to a real issue in the real world)

**Comment by bkchr** [comment]
I know how to solve it. The problem is more that it requires a new version of the host functions to proper version it. I can write a little bit more tomorrow.

**Comment by crystalin** [comment]
@bkchr if that helps, I tested #6230 on Moonbeam and there is a slight increase in PoV in some cases as expected but nothign too big

**Comment by bkchr** [comment]
Okay after some more thinking, I think I came to some conclusion on how to handle this.

We will need to version the `storage_proof_size` host function by introducing a second version. This is required because we will change the way the proof size is calculated while building the block and this is a breaking change.

Between calls to `storage_proof_size` we will need to collect `set_storage` calls, aka everything that will change the state. When the function is called we will need to "commit" these changes to the trie. This way we are already doing what `storage_root` would do at the end of the block. This intermediate trie, we are building while commiting the changes, should be preserved. This way we can directly reuse at the end of the block and not need to redo the operation again. The collection of operations that need to be committed on calling `storage_proof_size` needs to take storage transactions into account, this means that operations may get removed from this list again because the transaction was reverted. 
We also need to make the intermediate trie "storage transaction resilient". While `storage_proof_size` is always only called after a transaction got applied and the likelihood of it getting called as part of a storage transaction is low, we still need to account for that. This means that we need to undo certain operations to the trie when a `storage_proof_size` call was part of a storage transaction that got removed.

(Please don't get confused by `storage transaction` and `transaction`, both are different things :P The latter I mean a transaction from the user that is applied to a block)

**Comment by RomarQ** [comment]
Hi @bkchr 

Is this blocking for completly remove the `full-proof-size` feature? [Add feature to allow Aura collator to use full PoV size](https://github.com/paritytech/polkadot-sdk/pull/5393#issuecomment-2309770186)

**Comment by bkchr** [comment]
> Is this blocking for completly remove the `full-pov-size` feature?

AFAIK (and hopefully) yes

**Comment by RomarQ** [comment]
> > Is this blocking for completely remove the `full-pov-size` feature?
> 
> AFAIK (and hopefully) yes

Thanks for conforming. At moonbeam, we will just enable `full-pov-size` on staging networks for now.

**Comment by bkchr** [comment]
Yeah that works, but be aware that depending on the access pattern, you may produce too big PoVs.

**Comment by bkchr** [comment]
After this issue is done, we probably can revert: https://github.com/paritytech/polkadot-sdk/pull/8040

**Comment by bkchr** [comment]
@skunert and me have discussed this recently in the context of transaction streaming. There we will need to have the intermediate storage root after each transaction to forward it to light clients etc. If we would calculate the intermediate storage root after each transaction, we would also solve our issue at hand here. We for sure need to be a little bit smarter on how we calculate the storage root. Aka not pushing the entire overlay to the storage root calculation every time and only the stuff that changed. 

The good thing of doing this via storage root call, we can fix it entirely from the runtime side. Aka we don't need to take any care on how to version this as this would be only used when the runtime does the call. 

One of the big problems would be when to call `storage_root`. We need to call it before calling `proof_size`, but as we would still do some changes to the state after calling `proof_size`, the storage root at this point in time would not contain all the changes of the transaction. One obvious thing that is written afterwards is the actual used proof size and also the `AppliedExtrinsic` event. So, it would not be that useful for the transaction streaming. But, as the changes between the `storage_root` call before `proof_size` and a potential second call to `storage_root` in the block production logic would be quite small, the calculation should be quite fast. Also most of the changes should only be about changing existing nodes and thus all the intermediate nodes to the top, but no inserting/deletion should be required.

All in all, I think the `storage_root` solution sounds like a good solution, but we should before do some benchmarks and for sure we will need some optimizations to `storage_root` and the entire handling of it in the node. But none of them should be breaking changes.

**Comment by skunert** [comment]
EDIT: Concerns not relevant, we recently discussed this topic again. My original impression was that we will load more values from the trie by calculating the storage root in between. But that is actually not the case. Temporary values will be deleted in their pallets `on_finalize` hook, therefore `None` lands in the overlay and the path would be loaded from the OG trie anyway. 

~I think its worth to mention one side effect of the work here.~

~In FRAME, we sometimes see the pattern of writing temporary values. For example, a pallet might write temporary values in `on_initialize` and kill the corresponding storage items again in `on_finalize`. In the past, such temporary items did not have an impact on the PoV size, since these temporary items were never committed to the trie.~

~After the accuracy fixes in this PR however, these temporary values will have an impact. Since we will basically commit values mid-block, trie nodes on the path to the temporary values will be accessed. The impact on the PoV size will for sure depend on the pallet and exact storage patterns.~

~My impression is that the impact will be low, given the following assumptions:~
- ~The temporary values themselves will not become part of the PoV. If your pallet writes 2MB in `on_initialize` and deletes them in `on_finalize`, only the trie-nodes on the path are added, as the value itself is not read from the trie.~
- ~Since the key for storage values has the same prefix for all items in a given pallet, I assume that many of the parent nodes will already be loaded anyway by accessing other storage in the pallet. I expect that the biggest negative impact will occur if _only_ the temporary values are accessed, for example when there is no other interaction with the pallet during a given block.~

**Comment by michalkucharczyk** [comment]
> This is working kind of good, but still misses to account for some nodes that are for example loaded when merging two nodes. 

~@bkchr would you please elaborate more what scenario you mean here? I am trying to reason about this and find the case when this could happen - but so far I failed. Would appreciate some more explanation on this. (So far I think this is related [`fix`](https://github.com/paritytech/trie/blob/master/trie-db/src/triedbmut.rs#L1535) function, but am not able to easily find the case where this (_this_ meaning `commit` triggers extra loads comparing to other ops) actually could happen (assuming no extensions are used)). (cc: @skunert )~

**Comment by michalkucharczyk** [comment]
Summary of work done so far: https://hackmd.io/@_FY3-hvwQZ6cX_4n8zYUNA/Sy86uf2Kel
